### PR TITLE
* Easier to configure

### PR DESCRIPTION
**Pull request by Boychenko**
* Poison Queue
* Better error handling
* Keeps message invisible during processing
* replaces inheritance with composition
* enables C# 8 nullable feature
* enables treat warning as error
* Adds telemetry

### CODE REVIEW

File: src/AzureStorageQueueWorker.cs [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+namespace Likvido.Worker.AzureStorageQueue
+{
+    public class AzureStorageQueueWorker<TMessage, TMessageProcessor> : BackgroundService
+        where TMessageProcessor : IMessageProcessor<TMessage>
+    {
+        private readonly ILogger _logger;
+        private readonly AzureStorageQueueWorkerOptions<TMessageProcessor> _workerOptions;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IHostApplicationLifetime _hostApplicationLifetime;
+        private readonly AsyncPolicy _deleteMessagePolcy;
```
Comment (spewu) [inline src/AzureStorageQueueWorker.cs]: _deleteMessagePolcy -> _deleteMessagePolicy

File: src/AzureStorageQueueWorker.cs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+        private readonly AsyncPolicy _updateMessagePolcy;
```
Comment (spewu) [inline src/AzureStorageQueueWorker.cs]: _updateMessagePolcy -> _updateMessagePolicy

File: src/AzureStorageQueueWorker.cs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        private readonly SemaphoreSlim _messageReceiptSemaphore = new SemaphoreSlim(1, 1);
+        private QueueClient _posionQueueClient = null;
```
Comment (spewu) [inline src/AzureStorageQueueWorker.cs]: _posionQueueClient -> _poisonQueueClient

File: src/AzureStorageQueueWorker.cs [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        private readonly TelemetryClient _telemetryClient;
+
+        private QueueClient PosionQueueClient
```
Comment (spewu) [inline src/AzureStorageQueueWorker.cs]: PosionQueueClient -> PoisonQueueClient

File: src/AzureStorageQueueWorker.cs [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+            {
+                //skip
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Update message visibility has failed.");
+            }
+        }
+
+        private AsyncPolicy GetDeleteMessagePolcy(string failureText)
```
Comment (spewu) [inline src/AzureStorageQueueWorker.cs]: GetDeleteMessagePolcy -> GetDeleteMessagePolicy

File: src/AzureStorageQueueWorker.cs [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+        {
+            var attemptsCount = 3;
+            //delay short be short otherwise we will block other messages to be processed
```
Comment (spewu) [inline src/AzureStorageQueueWorker.cs]: Can you rewrite this comment, please? - maybe `the delay has to be short otherwise we will block other messages from being processed`

File: src/AzureStorageQueueWorkerOptions.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            {
+                if (TimeSpan.FromSeconds(5) > value)
+                {
+                    throw new ArgumentException("Timeout must be 5 seconds minimum", nameof(VisibilityTimeout));
+                }
+                _visibilityTimeout = value;
+            }
+        }
+
+        [NotNull]
```
Comment (spewu) [inline src/AzureStorageQueueWorkerOptions.cs]: If you do not set `QueueName` when you create a new instance of `AzureStorageQueueWorkerOptions`, then it would be null, right? - so why are you marking it with this attribute? (most likely I am missing something here)

This also reminds me that we should enable the C# 8 nullable checks for this project: https://stackoverflow.com/a/56267555/310937

File: src/AzureStorageQueueWorkerOptionsBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Likvido.Worker.AzureStorageQueue
+{
+    public class AzureStorageQueueWorkerOptionsBuilder
```
Comment (spewu) [inline src/AzureStorageQueueWorkerOptionsBuilder.cs:3]: What is the usefulness of this class?

File: src/AzureStorageQueueWorkerServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        private static AzureStorageQueueWorkerOptions<TMessageProcessor> CreateAzureStorageQueueWorkerOptions<TMessageProcessor>(
+            [NotNull] IServiceProvider applicationServiceProvider,
+            [NotNull] Action<IServiceProvider, AzureStorageQueueWorkerOptionsBuilder> optionsAction)
+        {
+            var options = new AzureStorageQueueWorkerOptions<TMessageProcessor>();
+            var builder = new AzureStorageQueueWorkerOptionsBuilder(options);
+
+            optionsAction?.Invoke(applicationServiceProvider, builder);
+
+            return options;
```
Comment (spewu) [inline src/AzureStorageQueueWorkerServiceCollectionExtensions.cs:75]: We should verify that both `QueueName` & `AzureStorageConnectionString` has been provided, and throw an exception if not

Comment (spewu) [reply src/AzureStorageQueueWorkerServiceCollectionExtensions.cs:75]: Alternatively, move them to the `AddMessageProcessor` extension method, to force setting them

File: README.md [hunk 10]
```diff
@@ -4,24 +4,31 @@
...
 }
 ```
+Register created class in Program:
+```
+services.AddMessageProcessor<MyCustomMessage, MessageProcessor>(
+    (sp, b) =>
+    {
+        b.Options.AzureStorageConnectionString = "";
+        b.Options.QueueName = "";
+    });
```
Comment (spewu) [inline README.md:33]: Maybe also add a note about enabling application insights telemetry, now that we have taken a dependency on it:

`services.AddApplicationInsightsTelemetryWorkerService();`

File: src/QueueClientExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+            }
+            catch (RequestFailedException exception)
+            when (exception.ErrorCode == QueueErrorCode.QueueNotFound
+               && exception.Status == 404)
+            {
+                //swallow
+            }
+
+            await queueClient.CreateIfNotExistsAsync(cancellationToken: cancellationToken);
+            await queueClient.SendMessageAsync(message, cancellationToken);
```
Comment (spewu) [inline src/QueueClientExtensions.cs]: This is a matter of preference, but I would probably move these two lines into the `catch`, and then remove the `return` from line 28

File: src/QueueMessageExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Text;
+using Azure.Storage.Queues.Models;
+
+namespace Likvido.Worker.AzureStorageQueue
+{
+    public static class QueueMessageExtensions
+    {
+        public static string EncocodeMessageText(this string message)
```
Comment (spewu) [inline src/QueueMessageExtensions.cs:9]: Is this method used anywhere?

File: src/AzureStorageQueueWorker.cs [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+                            KeepMessageInvisibleAsync(queueClient, messageDetails, visibilityToken), visibilityToken);
+
+                        scope = _serviceProvider.CreateScope();
+                        var messageProcessor = scope.ServiceProvider.GetRequiredService<TMessageProcessor>();
+                        await messageProcessor.ProcessMessage(GetTypedMessage(queueMessage), stoppingToken);
+                        processed = true;
+
+                        //stoppingToken aren't passed here. We should try to delete message even if cancellation was requested
+                        //if operation was done
+                        await DeleteMessageAsync(queueClient, messageDetails, updateMessageTokenSource);
```
Comment (spewu) [inline src/AzureStorageQueueWorker.cs:111]: Won't `updateMessageTokenSource` be cancelled whenever `stoppingToken` gets cancelled?

**Review by spewu — CHANGES_REQUESTED** [review]
Overall looks really nice!

File: src/AzureStorageQueueWorkerOptions.cs (hunk 7 above)
Comment (Boychenko) [reply src/AzureStorageQueueWorkerOptions.cs]: 1. this attribute shows that the value must be provided. When you are consuming the library you can hit F12 and you'll this attribute.
2. I'll give nullable checks a try. Makes a lot of sense

File: src/AzureStorageQueueWorkerOptionsBuilder.cs (hunk 8 above)
Comment (Boychenko) [reply src/AzureStorageQueueWorkerOptionsBuilder.cs:3]: Mostly for future purposes, we can add more methods with extra more complicated validation instead of having all validation in the options. Keeping backward compatibility will be easier

File: src/AzureStorageQueueWorkerServiceCollectionExtensions.cs (hunk 9 above)
Comment (Boychenko) [reply src/AzureStorageQueueWorkerServiceCollectionExtensions.cs:75]: yep, good point
I'm still not sure. The way it works right now. This two lines will fail if anything is wrong with `AzureStorageConnectionString` or `QueueName`
https://github.com/Likvido/Likvido.Worker.AzureStorageQueue/blob/4f8a4b836b2d4aa72f13934ed3d339a4c5d60fb2/src/AzureStorageQueueWorker.cs#L66-L67

and in the following code, a consumer can control if should swallow the error or stop with a particular exit code using `_workerOptions.SetupIssueStopHostCod`
https://github.com/Likvido/Likvido.Worker.AzureStorageQueue/blob/4f8a4b836b2d4aa72f13934ed3d339a4c5d60fb2/src/AzureStorageQueueWorker.cs#L135-L138

But if you want I can add throwing an exception during init config phase but this will prevent the whole app from starting

File: README.md (hunk 10 above)
Comment (Boychenko) [reply README.md:33]: but default it'll be enabled. I'll add a note that it can be disabled via param but then it should be configured in the consumption code

File: src/QueueMessageExtensions.cs (hunk 12 above)
Comment (Boychenko) [reply src/QueueMessageExtensions.cs:9]: I've added this method to be used via consumers. To match `GetDecodedMessageText`

File: src/QueueClientExtensions.cs (hunk 11 above)
Comment (Boychenko) [reply src/QueueClientExtensions.cs]: Initially, I've written it using your approach but then found this version in the azure-sdk and since it less nested it probably a bit easier to read but I don't have a strong opinion on this

File: src/AzureStorageQueueWorker.cs (hunk 13 above)
Comment (Boychenko) [reply src/AzureStorageQueueWorker.cs:111]: yes, `updateMessageTokenSource.Token` will become cancelled. But `updateMessageTokenSource` is passed to cancel `UpdateVisibility` logic on message deletion

File: src/AzureStorageQueueWorkerOptions.cs (hunk 7 above)
Comment (spewu) [reply src/AzureStorageQueueWorkerOptions.cs]: ok, so if you do:

```
var options = new AzureStorageQueueWorkerOptions();
var queueName = options.QueueName;
```

will you get a compiler warning or? (because in this case you have not set `QueueName`)

From what I can read about the NotNullAttribute, it just tells a consumer of the property that the value will never be null (even though the type allows null) ... but in my example above, the `queueName` would be null

File: README.md (hunk 10 above)
Comment (spewu) [reply README.md:33]: Ahh, sorry - I missed that you enable it inside `AzureStorageQueueWorkerServiceCollectionExtensions` 👍

File: src/AzureStorageQueueWorkerOptionsBuilder.cs (hunk 8 above)
Comment (spewu) [reply src/AzureStorageQueueWorkerOptionsBuilder.cs:3]: hmm... ok then

File: src/AzureStorageQueueWorkerServiceCollectionExtensions.cs (hunk 9 above)
Comment (spewu) [reply src/AzureStorageQueueWorkerServiceCollectionExtensions.cs:75]: I would prefer throwing the exception during initialization

File: src/QueueClientExtensions.cs (hunk 11 above)
Comment (spewu) [reply src/QueueClientExtensions.cs]: I think the level of nesting here is so small that it is easy to grasp. As I said, this comes down to opinion/preference, and in this case I prefer getting rid of the return statement, as I think it can be easily missed when reading the code

File: src/QueueMessageExtensions.cs (hunk 12 above)
Comment (spewu) [reply src/QueueMessageExtensions.cs:9]: You mean it should be used by producers, right?

So, the idea is that you base64 encode the message before pushing it to a queue?

File: src/AzureStorageQueueWorker.cs (hunk 13 above)
Comment (spewu) [reply src/AzureStorageQueueWorker.cs:111]: ahh, yea - makes sense

File: src/QueueMessageExtensions.cs (hunk 12 above)
Comment (Boychenko) [reply src/QueueMessageExtensions.cs:9]: Yes, in long term I'd say we should create another package something like "Likvido.Utils"

File: src/AzureStorageQueueWorkerOptions.cs (hunk 7 above)
Comment (Boychenko) [reply src/AzureStorageQueueWorkerOptions.cs]: I considered this 
https://stackoverflow.com/a/32806406

Comment (spewu) [reply src/AzureStorageQueueWorkerOptions.cs]: Ok, I don't really see the value of that attribute. I think we should remove it, and instead enable C# 8 nullable checks

File: src/AzureStorageQueueWorker.cs [hunk 14]
```diff
@@ -63,20 +63,31 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
             using var logScope = _logger.BeginScope("{Processor} reads {queueName}", processorName, _workerOptions.QueueName);
             try
             {
+                await Task.Yield();
```
Comment (spewu) [inline src/AzureStorageQueueWorker.cs]: Why is this needed?

File: src/AzureStorageQueueWorker.cs [hunk 15]
```diff
@@ -63,20 +63,31 @@ protected override async Task ExecuteAsync(CancellationToken stoppingToken)
...
                     try
                     {
-                        var queueMessageResponse = await queueClient.ReceiveMessagesAsync(1, _workerOptions.VisibilityTimeout, stoppingToken);
-                        queueMessage = queueMessageResponse.Value.FirstOrDefault();
+                        //clean values here before any code
+                        //otherwisewe can get previous loop cycle value in the catch/finally
+                        queueMessage = null;
+                        messageDetails = null;
+                        updateQueueMessageTask = null;
+                        updateMessageTokenSource = null;
```
Comment (spewu) [inline src/AzureStorageQueueWorker.cs]: Can you elaborate on this? - I don't understand how getting the previous loop value is possible since these variables are all defined at the beginning of the loop

File: src/LoggerExtensions.cs [hunk 16]
```diff
@@ -1,14 +1,41 @@
...
         }
 
         public static void OperationCancelledExceptionOccurred(this ILogger logger) => logger.Log(LogLevel.Information, EventIds.OperationCancelledExceptionCaught, "A task/operation cancelled exception was caught.");
+
+        public static void UnhandledMessageProcessingExceptionOccurred(this ILogger logger, Exception ex)
+            => logger.Log(
+                LogLevel.Error,
+                EventIds.UnhandledMessageProcessingException,
+                ex,
+                "An unhandled exception during messages processing was caught.");
```
Comment (spewu) [inline src/LoggerExtensions.cs]: An unhandled exception during message processing was caught.

File: src/LoggerExtensions.cs [hunk 17]
```diff
@@ -1,14 +1,41 @@
...
+                ex,
+                message ?? "An unknown setup exception was caught.",
+                args);
+
+        public static void CustomErrorHandlerExceptionOccured(this ILogger logger, Exception ex)
+            => logger.Log(
+                LogLevel.Error,
+                EventIds.CustomErrorHandlerException,
+                ex,
+                "An unhandled exception was thrown on custom exception hander call.");
```
Comment (spewu) [inline src/LoggerExtensions.cs]: An unhandled exception was thrown on a custom exception handler call.

**Review by spewu — CHANGES_REQUESTED** [review]

File: src/AzureStorageQueueWorker.cs (hunk 14 above)
Comment (Boychenko) [reply src/AzureStorageQueueWorker.cs]: test code, sorry

File: src/AzureStorageQueueWorker.cs (hunk 15 above)
Comment (Boychenko) [reply src/AzureStorageQueueWorker.cs]: if the following operation fails on the second or following loops after the successful one
https://github.com/Likvido/Likvido.Worker.AzureStorageQueue/blob/dbaf813691823e9baa1153e91b4ab55bc7f1fff1/src/AzureStorageQueueWorker.cs#L87-L88

Comment (spewu) [reply src/AzureStorageQueueWorker.cs]: The loop starts on line 69, and all of those variables are defined and initialized on lines 72 - 75 ... so they would be null - even on second loop. 

There must be a detail I am missing?

Comment (Boychenko) [reply src/AzureStorageQueueWorker.cs]: yeah, you are right :)
I'll fix this shortly
some leftovers after refactoring

**Review by spewu — APPROVED** [review]
👍
